PKI/TLS improvements (#13622)
cockroachdb/docs: v22.1/cockroach-cert.md

summary: A secure CockroachDB cluster uses TLS for encrypted inter-node and client-node communication.
toc: true
key: create-security-certificates.html
filter_category: security_cert
filter_html: Use cockroach cert
filter_sort: 1
docs_area: reference.cli
---

> To secure your CockroachDB cluster's inter-node and client-node communication, you need to provide a Certificate Authority (CA) certificate that has been used to sign keys and certificates (SSLs) for:

Reviewer: ```suggestion
To secure your CockroachDB cluster's communication between nodes and between clients and nodes, you need to provide a Certificate Authority (CA) certificate that has been used to sign keys and certificates (SSLs) for:
```
Just a suggestion that might add more clarity.

- Nodes
- Clients
- DB Console (optional)

> To create these certificates and keys, use the `cockroach cert` [commands](cockroach-commands.html) with the appropriate subcommands and flags, use [`openssl` commands](https://wiki.openssl.org/index.php/), or use a [custom CA](create-security-certificates-custom-ca.html) (for example, a public CA or your organizational CA).

Reviewer: ```suggestion
To create these certificates and keys, choose **one** of the following options:

- Use the `cockroach cert` [commands](cockroach-commands.html) with the appropriate subcommands and flags.
- Use [`openssl` commands](https://wiki.openssl.org/index.php/).
- Use a [custom CA](create-security-certificates-custom-ca.html) (for example, a public CA or your organizational CA).
```

This is a really complex sentence that feels like a run-on even though it isn't. Turning it into bullets might make it more readable.

{% include filter-tabs.md %}

> {{site.data.alerts.callout_success}}For details about when and how to change security certificates without restarting nodes, see <a href="rotate-certificates.html">Rotate Security Certificates</a>.{{site.data.alerts.end}}

Reviewer: Curious why you didn't use a Markdown link here.

## How security certificates work

> 1. Using the `cockroach cert` command, you create a CA certificate and key and then node and client certificates that are signed by the CA certificate. Since you need access to a copy of the CA certificate and key to create node and client certs, it's best to create everything in one place.

Reviewer: In some places you say "certs" and others "certificates". It's probably best to use the whole word everywhere.

2. You then upload the appropriate node certificate and key and the CA certificate to each node, and you upload the appropriate client certificate and key and the CA certificate to each client.

> 3. When nodes establish contact to each other, and when clients establish contact to nodes, they use the CA certificate to verify each other's identity.

Reviewer: ```suggestion
3. When nodes establish connections with each other, and when clients establish connections with nodes, they use the CA certificate to verify each other's identity.
```

## Subcommands

Subcommand | Usage
-----------|------
`create-ca` | Create the self-signed certificate authority (CA), which you'll use to create and authenticate certificates for your entire cluster.
`create-node` | Create a certificate and key for a specific node in the cluster. You specify all addresses at which the node can be reached and pass appropriate flags.
`create-client` | Create a certificate and key for a [specific user](create-user.html) accessing the cluster from a client. You specify the username of the user who will use the certificate and pass appropriate flags.
`list` | List certificates and keys found in the certificate directory.

## Certificate directory

> When using `cockroach cert` to create node and client certificates, you will need access to a local copy of the CA certificate and key. It is therefore recommended to create all certificates and keys in one place and then distribute node and client certificates and keys appropriately. For the CA key, be sure to store it somewhere safe and keep a backup; if you lose it, you will not be able to add new nodes or clients to your cluster. For a tutorial of this process, see [Manual Deployment](manual-deployment.html).

Reviewer: ```suggestion
When using `cockroach cert` to create node and client certificates, you need access to a local copy of the CA certificate and key. For that reason, we recommend that you create all certificates and keys in one place and then distribute node and client certificates and keys appropriately. For the CA key, be sure to store it somewhere safe and keep a backup; if you lose it, you cannot add new nodes or clients to your cluster. For a tutorial of this process, see [Manual Deployment](manual-deployment.html).
```

summary: A secure CockroachDB cluster uses TLS for encrypted inter-node and client-node communication.
toc: true
key: create-security-certificates.html
docs_area: reference.cli
---

The CockroachDB CLI's `cockroach cert` command allows you to generate [private key/public certificate pairs for TLS authentication and encryption in communication between CockroachDB nodes, and from SQL clients to the cluster.

See:

- [Transport Layer Security (TLS) and Public Key Infrastructure (PKI)](security-reference/transport-layer-security.html)
- [Using the CockroachDB CLI to provision a development cluster](manage-certs-cli.html).
- [Using Google Cloud Platform to manage PKI certificates](manage-certs-gcloud.html).

{{site.data.alerts.callout_info}}

> The ability to rapidly and locally generate private key/public certificate pairs is handy for development, but careful management of security certificates is an essential component of cluster security, and performing these tasks with a cloud-native tool such as Google Cloud Platform's Certificate Authority Service (CAS) offers many security advantages.

Author: will address this in follow up ticket for this doc.

- This check can be disabled by setting the environment variable `COCKROACH_SKIP_KEY_PERMISSION_CHECK` to `true`.

## Synopsis

Create the CA certificate and key:

~~~ shell

> $ cockroach cert create-ca \

Reviewer: I think we use curly braces for replaceable text. https://github.com/cockroachdb/docs/blob/master/StyleGuide.md#placeholders
